stan/math/prim/scal/meta/operands_and_partials.hpp

```diff
@@ -0,0 +1,120 @@
+        // size is used just for reverse mode to calculate the amount of memory
+        // slots required to store the operands and partials.
+        int size() const { return 0; }
+      };
+
+      // Base class shared between fvar and var specializations of uni- and
+      // multivariate edges will be in the prim/scal|arr|mat files.
+      // Single param version - underlying Var is the same as Op, so there's
+      // just one.
+      template <typename ViewElt, typename Op>
+      class ops_partials_edge_single {
+      private:
+        ViewElt partial_;
+      public:
+        const Op& operand_;
+        broadcast_array<ViewElt> partials_;
+        explicit ops_partials_edge_single(const Op& op)
+          : partial_(0), operand_(op), partials_(partial_) {}
+        // dump_operands implemented in specialization
+        int size() const { return 1; }
+      };
+    }  // namespace detail
+
+    /**
+     * This class builds partial derivatives with respect to a set of
+     * operands. There are two reason for the generality of this
+     * class. The first is to handle vector and scalar arguments
+     * without needing to write additional code. The second is to use
+     * this class for writing probability distributions that handle
+     * primitives, reverse mode, and forward mode variables
+     * seamlessly.
+     *
+     * Conceptually, this class is used when we want to manually calculate
+     * the derivative of a function and store this manual result on the
+     * autodiff stack in a sort of "compressed" form. Think of it like an
+     * easy-to-use interface to rev/core/precomputed_gradients.
+     *
+     * This class now supports multivariate use-cases as well by
+     * exposing edge#_.partials_vec
+     *
+     * This base template is used when all operands are primitives
+     * and we don't want to calculate derivatives at all. So all
+     * Op1 - Op4 must be arithmetic primitives like int or double.
+     *
+     * @tparam Op1 type of the first operand
+     * @tparam Op2 type of the second operand
+     * @tparam Op3 type of the third operand
+     * @tparam Op4 type of the fourth operand
+     * @tparam T_return_type return type of the expression. This defaults
+     *   to a template metaprogram that calculates the scalar promotion of
+     *   Op1 -- Op4
+     */
+    template <typename Op1 = double, typename Op2 = double,
+              typename Op3 = double, typename Op4 = double,
+              typename T_return_type
+              = typename return_type<Op1, Op2, Op3, Op4>::type>
+    class operands_and_partials {
+    public:
+      explicit operands_and_partials(const Op1& op1) {}
+      operands_and_partials(const Op1& op1, const Op2& op2) {}
+      operands_and_partials(const Op1& op1, const Op2& op2, const Op3& op3) {}
+      operands_and_partials(const Op1& op1, const Op2& op2, const Op3& op3,
+                            const Op4& op4) {}
+      typedef double ViewElt;
```

Comment:
Ha. The reason I'd want it to be private is since we're using a concept and not object inheritance, whoever is looking at this as the concept may mistake `ViewElt` as being part of the concept. It's not (specializations don't actually specify it).